src/Markdown/Block.elm

```diff
@@ -41,48 +58,177 @@ In the simplest case, you can pass this directly to a renderer:
-    | Body (List Inline)
-    | Html String (List Attribute) (List Block)
-    | UnorderedListBlock
-        (List
-            { task : Maybe Bool
-            , body : List Inline
-            }
-        )
-    | OrderedListBlock Int (List (List Inline))
-    | CodeBlock Markdown.CodeBlock.CodeBlock
-    | ThematicBreak
+    = -- Container Blocks
+      HtmlBlock (Html Block)
+    | UnorderedList (List (ListItem Inline))
+    | OrderedList Int (List (List Inline))
     | BlockQuote (List Block)
+      -- Leaf Blocks With Inlines
+    | Heading HeadingLevel (List Inline)
+    | Paragraph (List Inline)
+      -- Table TODO  https://github.github.com/gfm/#tables-extension-
+      -- Leaf Blocks Without Inlines
+    | CodeBlock { body : String, language : Maybe String }
+    | ThematicBreak
+
+
+{-| The value for an unordered list item, which may contain a task.
+-}
+type ListItem children
+    = ListItem Task (List children)
+
+
+{-| A task (or no task), which may be contained in a ListItem.
+-}
+type Task
+    = NoTask
+    | IncompleteTask
+    | CompletedTask
+
+
+{-| Markdown limits headings to level 6 or less. HTML does this, too (`<h7>` is not supported by most browsers).
+This type represents the possible heading levels that a Markdown heading block may have.
+
+If you do a heading level above 6, the `#` characters will be treated as literal `#`'s.
+
+-}
+type HeadingLevel
+    = H1
+    | H2
+    | H3
+    | H4
+    | H5
+    | H6
+
+
+{-| An Inline block. Note that `HtmlInline`s can contain Blocks, not just nested `Inline`s.
+-}
+type Inline
+    = HtmlInline (Html Block)
+    | Link String (Maybe String) (List Inline)
+    | Image String (Maybe String) (List Inline)
+    | Emphasis (List Inline)
+    | Strong (List Inline)
+      -- Strikethrough TODO  https://github.github.com/gfm/#strikethrough-extension-
+    | CodeSpan String
+    | Text String
+    | HardLineBreak
+
+
+{-| A convenience function so that you don't have to write a big case statement if you need a heading level as an Int.
+-}
+headingLevelToInt : HeadingLevel -> Int
+headingLevelToInt headingLevel =
+    case headingLevel of
+        H1 ->
+            1
+
+        H2 ->
+            2
+
+        H3 ->
+            3
+
+        H4 ->
+            4
+
+        H5 ->
+            5
+
+        H6 ->
+            6
+
 
+{-| Extract the text from a list of inlines.
+
+    inlines : List (Inline)
+    inlines =
+        [ Text "Heading with "
+        , Emphasis 1
+            [ Text "emphasis" ]
+        ]
+
+    extractText inlines == "Heading with emphasis"
+
+    -- Original string: "Heading with *emphasis*"
 
-{-| Represents styled inline text. For example, a header can include links, emphasis, etc.
 -}
-type alias Inline =
-    { style : InlineStyle, string : String }
+extractInlineText : List Inline -> String
+extractInlineText inlines =
+    List.foldl extractTextHelp "" inlines
+
+
+extractTextHelp : Inline -> String -> String
+extractTextHelp inline text =
```

Comment:
Is there a chance that a List of Inlines could cause this corecursion to overflow? I'm not sure how extensive this could get in practice.

Reply:
I don't think it's possible, if that happened then wouldn't it mean that the actual data Inline structure itself would have to be too large to traverse? I don't believe it's possible for this function to traverse any more steps than the actual number of wrapped items within the `Inline` value itself.